File: longestPalindromeSubseq.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */
#include"cpptools.h"
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>
#include<queue>

using namespace std;

//#define testMod

#ifdef testMod
void test()
{

    
}
#endif

#ifndef testMod
class Solution {
public:
    int longestPalindrome(string word1, string word2) {
        return longestPalindromeSubseq(word1+word2);
    }

    int longestPalindromeSubseq(string s) {
        if(s.size()==0)
            return 0;
        table=vector<vector<int>>(s.size(),vector<int>(s.size(),0));
        this->s=s;
        return find(0,(int)s.size()-1);
    }

    int find(int begin,int end)
    {
        if(begin>end)
            return 0;
        if(begin==end)
            return 1;
        if(table[begin][end]!=0)
            return table[begin][end];
        int maxLen=max(find(begin+1,end),find(begin,end-1));
        if(s[begin]==s[end])
            maxLen=max(maxLen,find(begin+1,end-1)+2);
        table[begin][end]=maxLen;
        return maxLen;
    }

private:
    vector<vector<int>> table;
    string s;
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    string word1 = "aa";
    string word2 = "bb";
    cout<<sl.longestPalindrome(word1,word2)<<endl;
#endif 
    return 0;
}

File: longestPalindromeSubseq.py
from typing import List


class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        mem = [[0] * i for i in range(n)]

        def find(start, end):
            if start > end:
                return 0

            if start == end:
                return 1

            if mem[end][start] > 0:
                return mem[end][start]

            if s[start] == s[end]:
                return 2 + find(start + 1, end - 1)
            tmp = max(find(start, end - 1), find(start + 1, end))
            mem[end][start] = tmp

            return tmp

        return find(0, n - 1)


sl = Solution()
s = "bbbab"
s = 'abcccbac'
# s = "cbbd"
print(sl.longestPalindromeSubseq(s))
